src/main/java/net/neoforged/neoforge/client/pipeline/PipelineModifierStack.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.pipeline;
+
+import com.mojang.blaze3d.pipeline.RenderPipeline;
+import it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Locale;
+import java.util.Map;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+public final class PipelineModifierStack {
+    private final Deque<ResourceKey<PipelineModifier>> modifiers = new ArrayDeque<>();
+    private final Map<ResourceKey<PipelineModifier>, Map<RenderPipeline, RenderPipeline>> modifierTransformCache = new Reference2ReferenceOpenHashMap<>();
```

Comment:
The inner map can only work with identity comparisons. Is that intended?

Reply:
(We could in principle intern `RenderPipeline` instances. I guess there is no point if each transform changes the ID though.

Reply:
Yes, that's intentional. The end result would be the same if we used a non-reference map since `RenderPipeline` itself doesn't implement `hashCode()` and `equals()`.
I had considered interning the pipelines but discarded the idea in favor of better traceability.